File: csharp/Vernacula.Phonemizer/Languages/Arabic/Restore.cs
/**
 * Arabic short-vowel restoration — a SUPPLEMENT-ONLY pass that repairs the words the neural diacritizer
 * (diacritizer.ts) leaves under-voweled.
 * Ported from src/languages/arabic/restore.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Arabic;

public static class Restore
{
    private static readonly JsRe HARAKAT_G = JsRegex.Compile("[ً-ْٰ]", "g"); // tanwīn, short vowels, shadda, sukūn, dagger alif
    private static string StripHarakat(string w) => HARAKAT_G.Replace(w, "");

    /**
     * Is a diacritized word's g2p output a SKELETON — 0 vowels, or a run of ≥3 consecutive consonant
     * segments?
     */
    public static bool IsSkeleton(string diacritizedWord)
    {
        var segs = G2p.ToSegments(diacritizedWord);
        if (segs.Count == 0) return false;
        int vowels = 0,
            run = 0,
            maxRun = 0;
        foreach (var s in segs)
        {
            if (s.Vowel)
            {
                vowels++;
                run = 0;
            }
            else
            {
                run++;
                if (run > maxRun) maxRun = run;
            }
        }
        return vowels == 0 || maxRun >= 3;
    }

    private static readonly (string Key, string Voc)[] CLITICS =
    {
        ("وال", "وَال"), ("فال", "فَال"), ("بال", "بِال"), ("كال", "كَال"), ("لل", "لِل"),
        ("ال", "ال"), ("و", "وَ"), ("ف", "فَ"), ("ب", "بِ"), ("ك", "كَ"), ("ل", "لِ"), ("س", "سَ"),
    };
    private static readonly string[] SUFFIXES = { "ها", "هما", "هم", "هن", "كما", "كم", "كن", "نا", "ه", "ك", "ي" };

    private const string ARABIC_CONSONANTS = "ءآأؤإئابةتثجحخدذرزسشصضطظعغفقكلمنهوي";
    private const string LONG_CARRIERS = "اوىي"; // alif/wāw/alif-maqṣūra/yāʾ — mater lectionis (long vowel)

    /**
     * Epenthesis floor: vocalize a bare consonant SKELETON so an OOV word (no lexicon/clitic hit) is at least
     * SAYABLE.
     */
    private static string Epenthesize(string u)
    {
        var cs = Js.CodePoints(u);
        bool IsCons(string? c) =>
            c is not null && ARABIC_CONSONANTS.Contains(c, StringComparison.Ordinal)
                          && !LONG_CARRIERS.Contains(c, StringComparison.Ordinal);
        var outp = "";
        var i = 0;
        if (cs.Count > 3 && cs[0] == "ا" && cs[1] == "ل")
        {
            outp = "ال";
            i = 2;
        }
        while (i < cs.Count)
        {
            var c = cs[i];
            if (!IsCons(c))
            {
                outp += c;
                i++;
                continue;
            }
            outp += c;
            i++;
            if (i < cs.Count && LONG_CARRIERS.Contains(cs[i], StringComparison.Ordinal))
            {
                outp += cs[i];
                i++;
            }
            else outp += "َ"; // inserted fatḥa (short nucleus)
            var j = i;
            while (j < cs.Count && IsCons(cs[j])) j++;
            var runLen = j - i;
            var codaLen = j >= cs.Count ? runLen : Math.Max(0, runLen - 1);
            for (var k = 0; k < codaLen; k++) outp += cs[i + k];
            i += codaLen;
        }
        return outp;
    }

    /**
     * Build a candidate VOCALIZED form for an undiacritized word: direct lexicon hit, else clitic-strip →
     * stem lookup → re-attach, else suffix-strip, else clitic+suffix, else the epenthesis floor.
     */
    private static string? BuildRestoredText(string word, IReadOnlyDictionary<string, string> lexicon)
    {
        var u = StripHarakat(word);
        if (u.Length < 2) return null;
        if (lexicon.TryGetValue(u, out var direct)) return direct;
        foreach (var (c, voc) in CLITICS)
        {
            if (u.Length > c.Length + 1 && u.StartsWith(c, StringComparison.Ordinal))
            {
                if (lexicon.TryGetValue(u[c.Length..], out var stem)) return voc + stem;
            }
        }
        foreach (var sfx in SUFFIXES)
        {
            if (u.Length > sfx.Length + 1 && u.EndsWith(sfx, StringComparison.Ordinal))
            {
                if (lexicon.TryGetValue(u[..^sfx.Length], out var stem)) return stem + sfx;
            }
        }
        foreach (var (c, voc) in CLITICS)
        {
            if (u.StartsWith(c, StringComparison.Ordinal))
            {
                var rest = u[c.Length..];
                foreach (var sfx in SUFFIXES)
                {
                    if (rest.Length > sfx.Length + 1 && rest.EndsWith(sfx, StringComparison.Ordinal))
                    {
                        if (lexicon.TryGetValue(rest[..^sfx.Length], out var stem)) return voc + stem + sfx;
                    }
                }
            }
        }
        return Epenthesize(u); // floor: sayable OOV
    }

    private static readonly JsRe WORD = JsRegex.Compile("[ء-يٰٱً-ْـ]+", "gu");

    /**
     * The supplement pass: for each Arabic word in the diacritizer's output whose g2p is a SKELETON, override
     * its vocalization from the lexicon (or the epenthesis floor) — but only if the override is itself no
     * longer a skeleton. Every non-skeleton word is left exactly as the diacritizer produced it.
     */
    public static string RestoreSkeletons(string vocalized, IReadOnlyDictionary<string, string> lexicon)
    {
        return WORD.Replace(vocalized, m =>
        {
            var w = m.Value;
            if (!IsSkeleton(w)) return w;
            var cand = BuildRestoredText(w, lexicon);
            if (cand is not null && !IsSkeleton(cand)) return cand;
            return w;
        });
    }

    /**
     * LEXICON-PRIMARY restoration: a direct Tashkeela lexicon hit is AUTHORITATIVE — it overrides the neural
     * diacritizer for any covered word. An OOV word keeps the neural output unless that output is a
     * skeleton, which then falls to the clitic/suffix strip plus the epenthesis floor.
     */
    public static string LexiconPrimary(string vocalized, IReadOnlyDictionary<string, string> lexicon)
    {
        return WORD.Replace(vocalized, m =>
        {
            var w = m.Value;
            if (lexicon.TryGetValue(StripHarakat(w), out var direct)) return direct;
            if (!IsSkeleton(w)) return w;
            var cand = BuildRestoredText(w, lexicon);
            return cand is not null && !IsSkeleton(cand) ? cand : w;
        });
    }
}

File: src/languages/arabic/restore.ts
/**
 * Arabic short-vowel restoration — a SUPPLEMENT-ONLY pass that repairs the words the neural diacritizer
 * (diacritizer.ts) leaves under-voweled. The BiLSTM is trained on running text and hedges to sukun on isolated /
 * OOV words, producing unpronounceable skeletons (يقول → jqwl, not jaquːl). This pass runs AFTER the diacritizer:
 * for each word whose g2p output is still a SKELETON (0 vowels, or a run of ≥3 consecutive consonants — Arabic
 * forbids CCC, so a 3-run marks a dropped short vowel), it overrides that word's vocalization from a Tashkeela-
 * derived PAUSAL lexicon (direct hit, else clitic/suffix strip → stem lookup), falling back to a syllable-aware
 * epenthesis FLOOR so the output is always sayable. It NEVER touches a word the diacritizer already voweled — so
 * it can only improve, never degrade.
 */

import { toSegments } from "./g2p.ts";
import { MANIFEST } from "./manifest.ts";

const HARAKAT_G = /[ً-ْٰ]/g; // tanwīn, short vowels, shadda, sukūn, dagger alif
const stripHarakat = (w: string): string => w.replace(HARAKAT_G, "");

/**
 * Is a diacritized word's g2p output a SKELETON — 0 vowels, or a run of ≥3 consecutive consonant segments? Uses
 * the engine's own segmentation (Seg.vowel), so no IPA-vowel guessing and no digraph miscounting. This is the
 * gate: the restoration fires ONLY on skeletons, never on an already-sayable word.
 */
export function isSkeleton(diacritizedWord: string): boolean {
    const segs = toSegments(diacritizedWord);
    if (segs.length === 0) return false;
    let vowels = 0,
        run = 0,
        maxRun = 0;
    for (const s of segs) {
        if (s.vowel) {
            vowels++;
            run = 0;
        } else {
            run++;
            if (run > maxRun) maxRun = run;
        }
    }
    return vowels === 0 || maxRun >= 3;
}

// Leading proclitics (undiacritized key → VOCALIZED form to re-attach), longest-first. The vocalized form gives
// the proclitic its own short vowel so the re-attached clitic isn't itself a skeleton. Bare article ال stays
// undiacritized — the g2p adds ʔal-.
const CLITICS: ReadonlyArray<readonly [string, string]> = [
    ["وال", "وَال"], ["فال", "فَال"], ["بال", "بِال"], ["كال", "كَال"], ["لل", "لِل"],
    ["ال", "ال"], ["و", "وَ"], ["ف", "فَ"], ["ب", "بِ"], ["ك", "كَ"], ["ل", "لِ"], ["س", "سَ"],
];
// Pronominal/object suffixes (undiacritized), re-attached after the voweled stem.
const SUFFIXES: readonly string[] = ["ها", "هما", "هم", "هن", "كما", "كم", "كن", "نا", "ه", "ك", "ي"];

// ⚠ BOTH FROM THE MANIFEST, which already held every letter in them. The consonant string additionally
// takes آ, ة and ا, which are vowel carriers rather than `consonants` entries — the delta is stated here
// rather than by re-spelling the whole inventory, so adding a consonant to arabic.jsonc reaches this scan.
// (Deriving it also admits the Perso-Arabic پچژڤڭگݣ, which the literal omitted; 0 of 4,531 async probes
// moved, because the 258,784-entry restoration lexicon holds no word containing one.)
const L = MANIFEST.letters;
const ARABIC_CONSONANTS = Object.keys(MANIFEST.consonants).join("") + L.alifMadda + L.taaMarbuta + L.alif;
/** alif / wāw / alif-maqṣūra / yāʾ — mater lectionis (long vowel). */
const LONG_CARRIERS = L.alif + L.waw + L.alifMaqsura + L.ya;

/**
 * Epenthesis floor: vocalize a bare consonant SKELETON so an OOV word (no lexicon/clitic hit) is at least SAYABLE.
 * SYLLABLE-AWARE (Arabic (C)V(C)(C)): each onset consonant gets ONE nucleus (an inserted fatḥa, or a following
 * long-vowel carrier that already supplies it); the consonants that follow stay as a coda cluster up to the next
 * onset — reading the "tight skeleton" (دنمارك → دَنمارك /danmaːrk/) rather than breaking every cluster.
 */
function epenthesize(u: string): string {
    const cs = [...u];
    const isCons = (c: string | undefined): boolean =>
        c !== undefined && ARABIC_CONSONANTS.includes(c) && !LONG_CARRIERS.includes(c);
    let out = "",
        i = 0;
    // leading definite article ال: pass through UNVOCALIZED so the g2p handles /ʔal-/ + sun-letter assimilation.
    if (cs[0] === "ا" && cs[1] === "ل" && cs.length > 3) {
        out = "ال";
        i = 2;
    }
    while (i < cs.length) {
        const c = cs[i]!;
        if (!isCons(c)) {
            out += c;
            i++;
            continue;
        }
        out += c;
        i++;
        if (i < cs.length && LONG_CARRIERS.includes(cs[i]!)) {
            out += cs[i]!;
            i++;
        } else out += "َ"; // inserted fatḥa (short nucleus)
        // coda cluster: the following consonant run MINUS its last member (which onsets the next syllable), unless
        // the run ends the word (all coda, CVCC).
        let j = i;
        while (isCons(cs[j])) j++;
        const runLen = j - i;
        const codaLen = j >= cs.length ? runLen : Math.max(0, runLen - 1);
        for (let k = 0; k < codaLen; k++) out += cs[i + k]!;
        i += codaLen;
    }
    return out;
}

/**
 * Build a candidate VOCALIZED form for an undiacritized word: direct lexicon hit, else clitic-strip → stem lookup
 * → re-attach, else suffix-strip, else clitic+suffix, else the epenthesis floor. Returns undefined for too-short
 * input.
 */
function buildRestoredText(word: string, lexicon: ReadonlyMap<string, string>): string | undefined {
    const u = stripHarakat(word);
    if (u.length < 2) return undefined;
    const direct = lexicon.get(u);
    if (direct !== undefined) return direct;
    for (const [c, voc] of CLITICS) {
        if (u.length > c.length + 1 && u.startsWith(c)) {
            const stem = lexicon.get(u.slice(c.length));
            if (stem !== undefined) return voc + stem;
        }
    }
    for (const s of SUFFIXES) {
        if (u.length > s.length + 1 && u.endsWith(s)) {
            const stem = lexicon.get(u.slice(0, -s.length));
            if (stem !== undefined) return stem + s;
        }
    }
    for (const [c, voc] of CLITICS) {
        if (u.startsWith(c)) {
            const rest = u.slice(c.length);
            for (const s of SUFFIXES) {
                if (rest.length > s.length + 1 && rest.endsWith(s)) {
                    const stem = lexicon.get(rest.slice(0, -s.length));
                    if (stem !== undefined) return voc + stem + s;
                }
            }
        }
    }
    return epenthesize(u); // floor: sayable OOV
}

// Arabic word token (letters + harakat) — everything else (spaces, digits, punctuation) passes through unchanged.
const WORD = /[ء-يٰٱً-ْـ]+/gu;

/**
 * The supplement pass: for each Arabic word in the diacritizer's output whose g2p is a SKELETON, override its
 * vocalization from the lexicon (or the epenthesis floor) — but keep the override only if it is itself NO LONGER a
 * skeleton (a bad derivation can't make things worse). Every non-skeleton word is left exactly as the diacritizer
 * produced it.
 */
export function restoreSkeletons(vocalized: string, lexicon: ReadonlyMap<string, string>): string {
    return vocalized.replace(WORD, (w) => {
        if (!isSkeleton(w)) return w;
        const cand = buildRestoredText(w, lexicon);
        if (cand !== undefined && !isSkeleton(cand)) return cand;
        return w;
    });
}

/**
 * LEXICON-PRIMARY restoration: a direct Tashkeela lexicon hit is AUTHORITATIVE — it overrides the neural
 * diacritizer for any covered word (the classical citation-form vocalization is right for isolated/dictionary
 * words, which the context-trained BiLSTM gets wrong out-of-distribution). An OOV word (no direct hit) keeps the
 * neural output, unless that output is a skeleton — then it falls to the clitic/suffix-strip + epenthesis floor.
 * (Trade-off vs restoreSkeletons: the lexicon's single reading wins even where the neural's CONTEXT would have
 * disambiguated mid-sentence — measured net-better on isolated/dictionary referees; running-text is the open Q.)
 */
export function lexiconPrimary(vocalized: string, lexicon: ReadonlyMap<string, string>): string {
    return vocalized.replace(WORD, (w) => {
        const direct = lexicon.get(stripHarakat(w));
        if (direct !== undefined) return direct;
        if (!isSkeleton(w)) return w;
        const cand = buildRestoredText(w, lexicon);
        return cand !== undefined && !isSkeleton(cand) ? cand : w;
    });
}
